torchrl/modules/models/multiagent.py

```diff
@@ -150,87 +243,52 @@ def __init__(
-                )
-                for _ in range(self.n_agents if not self.share_params else 1)
-            ]
+        super().__init__(
+            n_agents=n_agents,
+            centralised=centralised,
+            share_params=share_params,
+            device=device,
+            agent_dim=-2,
+            **kwargs,
         )
 
-    def forward(self, *inputs: Tuple[torch.Tensor]) -> torch.Tensor:
-        if len(inputs) > 1:
-            inputs = torch.cat([*inputs], -1)
-        else:
-            inputs = inputs[0]
-
-        if inputs.shape[-2:] != (self.n_agents, self.n_agent_inputs):
+    def _pre_forward_check(self, inputs):
+        if inputs.shape[-2] != self.n_agents:
             raise ValueError(
-                f"Multi-agent network expected input with last 2 dimensions {[self.n_agents, self.n_agent_inputs]},"
+                f"Multi-agent network expected input with shape[-2]={self.n_agents},"
                 f" but got {inputs.shape}"
             )
-
         # If the model is centralized, agents have full observability
         if self.centralised:
-            inputs = inputs.reshape(
-                *inputs.shape[:-2], self.n_agents * self.n_agent_inputs
-            )
-
-        # If parameters are not shared, each agent has its own network
-        if not self.share_params:
-            if self.centralised:
-                output = torch.stack(
-                    [net(inputs) for i, net in enumerate(self.agent_networks)],
-                    dim=-2,
-                )
-            else:
-                output = torch.stack(
-                    [
-                        net(inputs[..., i, :])
-                        for i, net in enumerate(self.agent_networks)
-                    ],
-                    dim=-2,
-                )
-        # If parameters are shared, agents use the same network
-        else:
-            output = self.agent_networks[0](inputs)
-
-            if self.centralised:
-                # If the parameters are shared, and it is centralised, all agents will have the same output
-                # We expand it to maintain the agent dimension, but values will be the same for all agents
-                output = output.view(*output.shape[:-1], self.n_agent_outputs)
-                output = output.unsqueeze(-2)
-                output = output.expand(
-                    *output.shape[:-2], self.n_agents, self.n_agent_outputs
-                )
-
-        if output.shape[-2:] != (self.n_agents, self.n_agent_outputs):
-            raise ValueError(
-                f"Multi-agent network expected output with last 2 dimensions {[self.n_agents, self.n_agent_outputs]},"
-                f" but got {output.shape}"
-            )
-
-        return output
+            inputs = inputs.flatten(-2, -1)
+        return inputs
+
+    def _build_single_net(self, *, device, **kwargs):
+        n_agent_inputs = self.n_agent_inputs
+        if self.centralised and n_agent_inputs is not None:
```

Comment:
Super cool that we are allowing `n_agent_inputs` to be `None` and thus have lazy modules.

We need to document this in the docstring and in the type hint